File: 1255-maximum-score-words-formed-by-letters/1255-maximum-score-words-formed-by-letters.py
class Solution:
    def maxScoreWords(self, words: List[str], letters: List[str], score: List[int]) -> int:
        
        def form_sub(idx, values = ''):
            if(idx >= len(words)):
                return get_score(values)
            
            return max(form_sub(idx + 1, values + words[idx]),\
                       form_sub(idx + 1, values))
        
        
        def get_score(word):
            letter_copy = Counter(letters)
            values = 0
            for c in word:
                if(not letter_copy[c]):
                    return 0

                letter_copy[c] -= 1
                values += score[ord(c) - 97]
                
            return values
        
        
        return form_sub(0)
                
                    
            

File: 1255-maximum-score-words-formed-by-letters/1255-maximum-score-words-formed-by-letters.go
func maxScoreWords(words []string, letters []byte, score []int) int {
    
    var getSubsetScore = func (subSetsWords []string, ) int {
        var lettersCount = make(map[byte]int)
        for _, l := range letters {
            lettersCount[l]++
        }
        
        var res = 0
        for _, word := range subSetsWords {
            var isFormable = true
            var currScore = 0

            for _, c := range word {
                currScore += score[c - 97]
                lettersCount[byte(c)]--
                if lettersCount[byte(c)] < 0 {
                    isFormable = false
                    break
                }
            }

            if isFormable {
                res += currScore
            }
        }
        return res
    }
    
    var subsets func(index int, curr []string) int
    subsets = func(index int, curr []string) int {
        if index >= len(words){
            return getSubsetScore(curr)
        }
        
        return max(subsets(index + 1, append(curr, words[index])), subsets(index + 1, curr))
    }
    
    return subsets(0, []string{})
}
